testsuite/tests/typing-objects-bugs/pr3968_bad.compilers.reference

```diff
@@ -14,33 +14,35 @@ Error: The class type
            val l :
              [ `Abs of
                  string *
-                 ([ `Abs of string * expr | `App of 'a * exp ] as 'b)
-             | `App of expr * expr ] as 'a
+                 ([> `App of
+                       [ `Abs of string * 'a | `App of expr * expr ] * exp ]
+                  as 'a)
+             | `App of expr * expr ]
```

Comment:
At first glance, the printed type has changed. Do you have any comments on what is happening in this file?

Reply:
Yes – the new type is actually more general, so this patch is *improving* this error.  To wit, observe that what is changing is the second argument to the outer `` `App``: it was

```ocaml
([ `Abs of string * expr | `App of 'a * exp ] as 'b)
```

where `'a` referred to the whole type and `'b` was unused.  Now, it is instead

```ocaml
([> `App of [ `Abs of string * 'a | `App of expr * expr ] * exp ] as 'a)
￼```

where `'a` is bound locally.  Recall from earlier in the input that `expr` is

```ocaml
type expr =
  [ `Abs of string * expr
  | `App of expr * expr
  ]
```

Thus, rather than requiring exactly `` `Abs`` and `` `App``, the new error message has a type that only requires `` `App``, and can contain more constructors; moreover, this new type allows the first argument to this `` `App`` to be less constrained; rather than requiring it to be the entire type (`'a`), it only mandates that the input `` `Abs`` contain a `string * 'a`.

(I wanted to come  up with a nice inclusion example you could paste here, but because the code doesn't compile, I couldn't quite figure out how to do so.)